src/main/kotlin/platform/mixin/inspection/injector/InvalidInjectorMethodSignatureInspection.kt

```diff
@@ -124,28 +120,44 @@ class InvalidInjectorMethodSignatureInspection : MixinInspection() {
         }
     }
 
-    private class ParametersQuickFix(private val expected: List<ParameterGroup>) : LocalQuickFix {
+    private class ParametersQuickFix(
+        private val expected: List<ParameterGroup>,
+        injectorType: InjectorType
+    ) : LocalQuickFix {
+
+        private val fixName = when (injectorType) {
+            InjectorType.INJECT -> "Fix method parameters"
+            else -> "Fix method parameters (won't keep captured locals)"
+        }
 
-        override fun getFamilyName() = "Fix method parameters"
+        override fun getFamilyName() = fixName
 
         override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
             val parameters = descriptor.psiElement as PsiParameterList
-            parameters.synchronize(
-                expected.flatMap {
-                    if (it.default) {
-                        it.parameters?.map {
-                            JavaPsiFacade.getElementFactory(project).createParameter(
-                                it.name ?: JavaCodeStyleManager.getInstance(project)
-                                    .suggestVariableName(VariableKind.PARAMETER, null, null, it.type).names
-                                    .firstOrNull() ?: "unknown",
-                                it.type
-                            )
-                        } ?: listOf()
-                    } else {
-                        listOf()
-                    }
+            // We want to preserve captured locals
+            val locals = parameters.parameters.dropWhile {
+                val type = it.type as? PsiClassType ?: return@dropWhile true
+                val fqname = type.fullQualifiedName ?: return@dropWhile true
+                return@dropWhile fqname != MixinConstants.Classes.CALLBACK_INFO &&
+                    fqname != MixinConstants.Classes.CALLBACK_INFO_RETURNABLE
+            }.drop(1) // the first element in the list is the CallbackInfo but we don't want it
```

Comment:
Should we check that this is in-fact the case or is this already guaranteed before we get to this point? (Thinking about incorrect code here).

As far as I can tell we actually only hit this point when the `checkParameters()` method fails, so I think a sufficieintly broken method could be missing the callback info.

I noticed in your unit tests all of the test cases do have this parameter.

Reply:
Good question, this list contains the locals, we only distinguish those thanks to the CallbackInfo so if this list has actual locals, it will also always contain the CallbackInfo.

If there's no CallbackInfo then what should be the locals are stripped as they are considered part of the mandatory parameters by the inspection.


I wanted to add more tests but I kinda (totally) forgot about it.

Reply:
That makes sense, thanks!